toc.md

```diff
@@ -428,77 +429,234 @@ An additional backlog of 936 hours of image processing jobs everyday, along with
+          Workerholic.manager.start
+        end
+      end
+
+      PIDS.freeze
+    end
+
+    # ...
+  end
+end
+```
+
+We fork processes if the user specified the option when starting Workerholic. Then we fork the number of processes specified minus one since the main process will serve as one of these processes, effectively having 1 parent process and N - 1 children processes. We make sure to store the processes ids `pid`, inside a `PIDS` constant for future reference. In each of these children processes we start the manager, which starts all the workers and the other components such as the Job Scheduler. Once these children processes are started we start the manager in the parent process.
+
+In the next section we will look at how forking children processes impacts the consumption of the available memory.
 
 ##### Processes and Memory consumption
-We're able to create more processes, but surely that comes with a cost like everything else right?
+
+A Ruby process can fork multiple children processes. The memory model for a process is as follows:
 
 ![efficiency_processes_memory_design](/images/efficiency_processes_memory_design.png)
 
-Before we get into that, let's talk a little bit more about processes. Each process has its own address space, stack, and heap. When you fork a process in Ruby, you create a child process which will get its own stack and share the same heap initially. This is called copy-on-write, meaning that the child process will share the same resources with the parent process, until modifications are made to that resource, in which case, that resource will be written into the child's own heap.
+Each process has its own address space, stack, and heap. When a process is forked in Ruby, a child process is created, which will get its own stack. As for the heap, depending on the OS the copy-on-write mechanism is employed:
+- children processes start by referencing the parent's process heap, effectively sharing the same resources with the parent process
+- as modifications to the resources on the heap occur, children processes write the mutated resources to their own heap, leaving the parent's heap unmodified
+
+We benchmarked the memory consumption of having one process and two processes by using a Rails application. Having one process takes up 125MB of memory, but having two processes don't take twice as much memory. This is the copy-on-write mechanism at work.
 
 ![memory_usage_processes](/images/memory_usage_processes.png)
 
-Here, we benchmarked both having one process and two processes. Having one process takes up 125MB of memory, but having two processes don't take twice as much memory. This is the copy-on-write mechanism at work.
+Let's assume the Rails application in our scenario is a pretty big Rails app and takes about 400MB. By forking 4 processes Workerholic, will at most take 1.6GB of memory out of 4GB, meaning, forking 4 processes is a viable option.
```

Comment:
Let's assume the Rails application in our scenario is a pretty big Rails app and takes about 400MB in memory. By forking 4 processes, Workerholic will at most take 1.6GB out of 4GB of RAM, meaning forking 4 processes is a viable option.